src/Security/Authentication/OAuth/src/OAuthHandler.cs

```diff
@@ -159,42 +159,10 @@ protected override async Task<HandleRequestResult> HandleRemoteAuthenticateAsync
             }
         }
 
-        protected virtual async Task<OAuthTokenResponse> ExchangeCodeAsync(string code, string redirectUri)
+        protected virtual async Task<OAuthTokenResponse> ExchangeCodeAsync(string code, string redirectUri, AuthenticationProperties properties)
```

Comment:
Is there a non-breaking way to add this new parameter? E.g. with a new overload?

Reply:
@Tratcher hi, I was trying to make this non-breaking but I really don't see how... An overload  won't work because now it is called from `HandleRemoteAuthenticateAsync` with the `properties` parameter. So to make, it non breakable I would need the new method to call the old one. And that is not possible because I would need to pass the `tokenRequestParameters` setted in the new method to the old one... I couldn't think of anything else, do you see a way out of this? How bad is this breaking change? n_n'

Reply:
https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/search?q=ExchangeCodeAsync&unscoped_q=ExchangeCodeAsync
It would break at least 9 providers, which is enough to try to mitigate. Let's think about it while we deal with the other issue.

Reply:
Ok, the first one about simply augmenting the Dictionary, I totally agree with you and I already changed it.
Will be giving some thought about how to not break it... But I don't see now how to make it without some smelly artifact...

Reply:
@Tratcher is it possible to add a new API ExchangeCodeAsnyc(ExchangeCodeContext context)?

This would future proof us a bit. We have used this in other places.

Reply:
@brentschmaltz yes, that's plausible. Either way it's a breaking change for many providers. If we added PKCE support directly we might be able to avoid this break. https://github.com/aspnet/AspNetCore/pull/9448#issuecomment-496999910